### ISSUE TITLE
Discuss testing strategy

### ISSUE THREAD

**Issue by subdavis** [girder/viime#281]
I've avoided writing tests to this point because things were very fluid and a large number of unit tests would simply have slowed us down and added no value.

I believe we're at a point now where a more explicit testing strategy is needed.  I don't think we should strive for % coverage, but we could probably identify areas in the code where tests would add value rather than simply adding maintenance burden.

In particular, I think it would be reasonable to test the vuex store, more of the things in utils, and some of the heavier logic in components such as DataTable, ProblemBar, Transform, and maybe Upload.

I don't have a strong desire to test what I consider to be "leaf nodes" of the application, such as individual analyses and the transform plots, but I'm open to suggestions on those.'

Testing vue components isn't an exact science, and we can follow the lead set by girder/girder_web_components where it makes sense.

**Comment by waxlamp** [comment]
> In particular, I think it would be reasonable to test the vuex store, more of the things in utils, and some of the heavier logic in components such as DataTable, ProblemBar, Transform, and maybe Upload.
> 

:100: 

I would also love to see us follow a more functional programming style and abstract out the testable bits into functions, even if those functions only exist in the files that call them. This would presumably make writing unit tests easier as well.

> I don't have a strong desire to test what I consider to be "leaf nodes" of the application, such as individual analyses and the transform plots, but I'm open to suggestions on those.'
> 

I think curating a set of data files that could either be automatically uploaded in a testing mode, or simply exist for us to manually play around with, would be a good start to a kind of integration testing. In particular, we already have `data_original.csv` and a mangled version of that file that requires masking, key column selection, header row selection, and imputation (or at least, I received such a mangled file one time from @subdavis). @jeffbaumes and I also created a simple test file with normally distributed columns of data for use in testing the PCA score and loadings plots.

**Comment by jeffbaumes** [comment]
I agree this should be done. Let's get the timing right. I expect that it will be a good idea to perform a feature freeze and hardening phase from Nov 1 to Nov 15 (MANA is Nov 15). Would that be a good time to introduce testing architecture and be guarding against regressions?

If the testing harness can unobstructively be added in a few hours, I'm fine with someone pursuing this anytme. I just don't want to enforce extensive testing on all PRs until after the MVP features are in (major one remaining is integration).

**Comment by subdavis** [comment]
Early November during feature freeze sounds like a great time to circle back to this.
